### PR TITLE
CDAP-3079 adding support for RecordWritable<StructuredRecord>

### PR DESCRIPTION
**Pull request by albertshau**
This change introduces support for StructuredRecord as the
type for a RecordWritable, similar to how RecordScannable
can have StructuredRecord as its type. Changed Table
so that it implements RecordWritable<StructuredRecord>,
making it possible to write to Tables through Hive queries.

Most of the work involves moving code from etl into cdap and
reusing it to deserialize json as a StructuredRecord, and to
convert a StructuredRecord into a Put.

Includes some cleanup and fixes around the ContextManager used in
explore to avoid reusing a closed context object.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT3059

**Comment by albertshau** [comment]
this is mostly moving code around and adding unit tests. In terms of functionality, the important parts are in ContextManager and DatasetOutputFormat.

File: cdap-explore/src/main/java/co/cask/cdap/hive/serde/ObjectSerializer.java [hunk 1]
```diff
@@ -72,7 +73,7 @@ public Writable serialize(Object o, ObjectInspector objectInspector) {
     List<TypeInfo> info = structTypeInfo.getAllStructFieldTypeInfos();
     List<String> names = structTypeInfo.getAllStructFieldNames();
 
-    Map<String, Object> recordMap = Maps.newConcurrentMap();
+    Map<String, Object> recordMap = new HashMap<>();
```
Comment (gsps1) [inline cdap-explore/src/main/java/co/cask/cdap/hive/serde/ObjectSerializer.java]: doesn't have to be ``concurrentMap`` ?

File: cdap-formats/src/main/java/co/cask/cdap/format/RecordPutTransformer.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+/**
+ * Transforms records into Puts.
+ */
+public class RecordPutTransformer {
+  private final String rowField;
+  private final Schema outputSchema;
+  
+  public RecordPutTransformer(String rowField, Schema outputSchema) {
+    if (outputSchema.getType() != Schema.Type.RECORD) {
+      throw new IllegalArgumentException("Schema must be a record instead of ''.");
```
Comment (gsps1) [inline cdap-formats/src/main/java/co/cask/cdap/format/RecordPutTransformer.java]: instead of missing the actual type..

File: cdap-explore/src/main/java/co/cask/cdap/hive/serde/ObjectSerializer.java (hunk 1 above)
Comment (albertshau) [inline cdap-explore/src/main/java/co/cask/cdap/hive/serde/ObjectSerializer.java]: don't know why it was before.  It is only used in this one method where it is serialized to json.  In fact, it can't be a concurrentMap because we need to support null values.

File: cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java [hunk 3]
```diff
@@ -78,11 +100,12 @@ public static Context getContext(@Nullable Configuration conf) throws IOExceptio
       // by the DatasetSerDe.initialize method, which needs to access the context when it wants to write to a
       // dataset, so it can cache its name. In that case, conf will be null, and it won't be possible to create a
       // context.
-      savedContext = createContext(conf);
+      return createContext(conf);
```
Comment (gsps1) [inline cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java]: there's comment above this line about why saving is important.. is that explanation not valid ? should we remove that then?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/AbstractTable.java [hunk 4]
```diff
@@ -192,6 +197,16 @@ public Type getRecordType() {
     return new StructuredRecordScanner(createSplitReader(split));
   }
 
+  @Override
+  public void write(StructuredRecord structuredRecord) throws IOException {
+    if (tableSchema == null || rowFieldName == null) {
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/AbstractTable.java]: can just check ``if (recordPutTransformer == null) {``

File: cdap-explore/src/main/java/co/cask/cdap/hive/datasets/DatasetOutputFormat.java [hunk 5]
```diff
@@ -67,11 +73,25 @@ public void checkOutputSpecs(FileSystem ignored, JobConf jobConf) throws IOExcep
...
       this.datasetAccessor = datasetAccessor;
       this.recordWritable = datasetAccessor.getDataset();
       this.recordType = recordWritable.getRecordType();
+      if (recordType == StructuredRecord.class) {
+        try {
+          DatasetSpecification datasetSpec = datasetAccessor.getDatasetSpec();
+          String schemaStr = datasetSpec.getProperty(DatasetProperties.SCHEMA);
+          // should never happen, as this should have been checked at table creation
+          if (schemaStr == null) {
+            throw new IllegalStateException("Dataset does not have the schema property.");
```
Comment (gsps1) [inline cdap-explore/src/main/java/co/cask/cdap/hive/datasets/DatasetOutputFormat.java]: maybe add the name of the dataset in the message by getting it from the spec.

File: cdap-explore/src/main/java/co/cask/cdap/hive/datasets/DatasetOutputFormat.java [hunk 6]
```diff
@@ -67,11 +73,25 @@ public void checkOutputSpecs(FileSystem ignored, JobConf jobConf) throws IOExcep
...
+          }
+          recordSchema = Schema.parseJson(schemaStr);
+        } catch (IOException | DatasetManagementException e) {
+          throw new RuntimeException("Unable to look up schema for dataset.", e);
```
Comment (gsps1) [inline cdap-explore/src/main/java/co/cask/cdap/hive/datasets/DatasetOutputFormat.java]: shouldn't we close ``recordWritable`` as there is an exception?

Comment (gsps1) [inline cdap-explore/src/main/java/co/cask/cdap/hive/datasets/DatasetOutputFormat.java]: same with the pattern where we getDatasetAccessor but don't close that on exception..

File: cdap-explore/src/test/java/co/cask/cdap/explore/service/datasets/EmailTableDefinition.java [hunk 7]
```diff
@@ -43,7 +43,7 @@
  * Dataset definition with a record scannable table, containing an extensive schema. Used for testing.
  */
 public class EmailTableDefinition extends AbstractDatasetDefinition<EmailTableDefinition.EmailTable, DatasetAdmin> {
-  private static final Schema SCHEMA = Schema.recordOf("email",
+  public static final Schema SCHEMA = Schema.recordOf("email",
```
Comment (gsps1) [inline cdap-explore/src/test/java/co/cask/cdap/explore/service/datasets/EmailTableDefinition.java]: why this change?

**Comment by gsps1** [comment]
@albertshau looked at the changes, i don't understand the changes in ContextManager and comments old/new about saving the context. 

but other than that, i looked at the other changes and just have few minor comments.

File: cdap-explore/src/test/java/co/cask/cdap/explore/service/datasets/EmailTableDefinition.java (hunk 7 above)
Comment (albertshau) [inline cdap-explore/src/test/java/co/cask/cdap/explore/service/datasets/EmailTableDefinition.java]: used in a unit test now that is in a different package.

**Comment by albertshau** [comment]
thanks for the review, I will try and add more detailed javadocs to ContextManager.  The class was very confusing to understand, so best to add as much explanation as possible now while it's still fresh in mind

File: cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java (hunk 3 above)
Comment (albertshau) [inline cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java]: good point, i will remove it.  The comment just seems wrong.

**Comment by sreevatsanraman** [comment]
@albertshau: Can this be merged?

**Comment by albertshau** [comment]
@gsps1 do you have any more comments?

**Comment by gsps1** [comment]
LGTM

**Comment by albertshau** [comment]
will merge if build is green:
http://builds.cask.co/browse/CDAP-DUT3059
